### ISSUE TITLE
Advertise SCT integration in 3D Slicer?  Replace FSLeyes script?

### ISSUE THREAD

**Issue by jcohenadad** [spinalcordtoolbox/spinalcordtoolbox#3430]
### Context

A collaboration with Michael Hardisty at University of Toronto, which started ~4 years ago, led to the development of an [SCT plugin](https://bitbucket.org/OrthopaedicBiomechanicsLab/sctwrappers/src/sct_docker/) for [3D Slicer](https://www.slicer.org/). 

We currently already have a GUI solution for SCT, which relies on FSLeyes. 

### Question

Should we join forces with Hardisty's lab and focus on the integration of SCT within Slicer?

Pros:
- Get help from Hardisty lab
- Slicer 3D is more Windows-friendly

Cons:
- Some users are attached to FSLeyes and do not use Slicer 3D.
- Slicer 3D is "heavy" to install and use (more complex than FSLeyes).

**Comment by joshuacwnewton** [comment]
> A collaboration with Michael Hardisty at University of Toronto, which started ~4 years ago, led to the development of an [SCT plugin](https://bitbucket.org/OrthopaedicBiomechanicsLab/sctwrappers/src/sct_docker/) for [3D Slicer](https://www.slicer.org/).

Skimming the repo, my first impression is that this project is much more involved than I would have expected. Here are some quick, surface-level observations:

* The SCT-Slicer integration appears to be bundled as part of a complete, standalone, custom Slicer installation, via [a 3.9GB direct-download link](https://queensuca-my.sharepoint.com/:u:/g/personal/15pyw_queensu_ca/EWTg9uKXDK1Bpe3eKZbx1HkB0wGQnk6fhv7JjRorxBQM3w?e=FncGPf). 
    * In other words, it doesn't seem possible to treat it as a light "plugin" that can be inserted into a fresh Slicer installation.
* The project contains individual wrapper "modules" corresponding to each SCT CLI script. 
     * Each wrapper module [duplicates a lot of the argument parsing code](https://bitbucket.org/OrthopaedicBiomechanicsLab/sctwrappers/src/385fd2227f54fc8610b388a9ceb61481b6d05406/SCTConcatTransfo/SCTConcatTransfo.py#lines-216), to the point where the wrapper modules feel quite heavy, almost like a second, separate set of SCT's CLI scripts.
     * Each wrapper module also [spins up an `sct_docker` image](https://bitbucket.org/OrthopaedicBiomechanicsLab/sctwrappers/src/385fd2227f54fc8610b388a9ceb61481b6d05406/SCTConcatTransfo/SCTConcatTransfo.py#lines-255) (!), which then accesses the SCT CLI script binaries via [a hardcoded path to SCT 4.2.1](https://bitbucket.org/OrthopaedicBiomechanicsLab/sctwrappers/src/385fd2227f54fc8610b388a9ceb61481b6d05406/SCTConcatTransfo/SCTConcatTransfo.py#lines-256) (!!!). And, as far as I know, we stopped maintaining [`sct_docker`](https://github.com/spinalcordtoolbox/sct_docker) due to maintenance costs (See: https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/2965).
     * So, unlike the FSLeyes script, it doesn't tap into a user's existing installation of SCT. Keeping the integration up to date, then, would require a lot of manual intervention.
* The documentation contains the quotes _"3D Slicer - We made a custom build with the integration of ANTs"_ and _"Slicer provides the GUI for Spinal Cord Tool Box (we used a custom version)"_. 
* I'm also nervous about the fact that this integration is kept in a separate repo, far from the core development of SCT.

> Should we join forces with Hardisty's lab and focus on the integration of SCT within Slicer?

Due to how much customization is involved here, the Slicer integration feels closer to a hybrid piece of software between SCT v4.2.1 and Slicer. Because of that, it seems _very_ difficult to update or maintain in its current state. So, I would have strong reservations about advertising the integration or directing our efforts towards it. 

(While I do also have reservations about the FSLeyes script (see https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3426#issuecomment-864204659), it has the benefit of being much closer to a true plugin that lightly wraps SCT. So, for the sake of maintainer workload, I would feel much more comfortable sticking with FSLeyes.)

**Comment by kousu** [comment]
I agree with Joshua. This project looks very hard to integrate with. They haven't integrated with SCT, they've vendored us. An integration would be something that finds an existing SCT installation and calls its API. If they've vendored us..well...they can do that! It's open source! But that means they're claiming SCT is *part of* their software, and I think that means they need to be responsible for it too. Especially since

We could join forces with them to help them get to a true integration by building an SCT conda-forge package and then using our expertise to help them build 3D Slicer conda package which lists it as a dependency.

---

Judging from e.g. https://bitbucket.org/OrthopaedicBiomechanicsLab/sctwrappers/src/sct_docker/SCTPropseg/SCTPropseg.py most of what they've done is added GUI input fields to wrap the parameters to SCT scripts:

![SCTPropseg GUI](https://bitbucket.org/OrthopaedicBiomechanicsLab/sctwrappers/raw/385fd2227f54fc8610b388a9ceb61481b6d05406/Screenshots/SCTPropseg.png)

I understand that there's a lot of desire among our users to have that UI. #3421 is a better path to that kind of GUI for us.

**Comment by jcohenadad** [comment]
I agree with your analysis @joshuacwnewton and @kousu. I am keen in reaching out to them and ask if they could consider using SCT via its APIs, instead of integrating it. We have invested quite a lot of time with their dev team already, and it looks like we will continue our collaboration in the near future.

**Comment by joshuacwnewton** [comment]
> I am keen in reaching out to them and ask if they could consider using SCT via its APIs, instead of integrating it.

I am hopeful that this this will become easier to do if SCT becomes pip-installable. Slicer's website mentions:

![Screenshot from 2021-06-24 16-58-22](https://user-images.githubusercontent.com/16181459/123331563-70e0bd80-d50d-11eb-8363-1a1074f18e85.png)

But, as I'm not too familiar with [development of Slicer extensions](https://www.slicer.org/wiki/Documentation/Nightly/Training#Developing_and_contributing_extensions_for_3D_Slicer), it's unclear to me if Slicer extensions can depend on pip packages. (See also: [Automatic install of Python packages](https://discourse.slicer.org/t/automatic-install-of-python-packages/7078), which describes a similar situation.)
